go/coredns/plugin/pfdns/pfdns.go

```diff
@@ -219,17 +221,26 @@ func (pf *pfdns) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
 			switch Type {
 			case "dnsenforcement", "inline":
 				var status = "unreg"
-				err = pf.Nodedb.QueryRow(mac, 1).Scan(&status)
+				var category string
+				err = pf.Nodedb.QueryRow(mac, 1).Scan(&status, &category)
 				if err != nil {
 					log.LoggerWContext(ctx).Error(fmt.Sprintf("error getting node status %s %s\n", mac, err))
 				}
 				// Defer to the proxy middleware if the device is registered
-				if status == "reg" && !violation {
+				if status == "reg" && !violation && category != "REJECT" {
 					log.LoggerWContext(ctx).Debug(srcIP + " : " + mac + " serve dns " + state.QName())
 					return pf.Next.ServeDNS(ctx, w, r)
 				}
+				if status == "reg" && category == "REJECT" {
+					rr, _ = dns.NewRR("30 IN A 127.0.0.1")
+					a.Answer = []dns.RR{rr}
+					log.LoggerWContext(ctx).Debug("REJECT " + mac + " IP " + srcIP + " Query " + state.QName())
+					state.SizeAndDo(a)
+					w.WriteMsg(a)
+					return 0, nil
+				}
 			}
-			answer, found := pf.DNSFilter.Get(state.QName())
+			answer, found := pf.DNSFilter.Get(mac + state.QName())
```

Comment:
I'd make the cache key, the mac, category, violation (as string) and the state.QName() and then the cache can be set to a high expiration since the cache key will accurately reflect the state of the endpoint.

Otherwise right now, if a MAC goes from rejected to allowed, its entries are cached for a minute while with the new cache key it misses the cache and returns the appropriate record

Also given you create the cache key in multiple places, I'd create a function to create the cache key from the parameters above.